linode/manager
# hotfix: DC-specific pricing -- monthly price decimals and invoice region column

## Description 📝
There are two issues with DC-specific pricing:
1. on past invoices, the Region column for invoice items is not backfilled due to how this is data is collected and populated on the back-end, which may cause confusion for users who have a region column with no data
2. monthly prices (previously all integers, displayed as `$n/mo`) are now displaying inconsistently -- most often with one decimal place, but in at least one instance (resize node pool drawer, 2GB linode) with way too many decimal places

## Major Changes 🔄
- Fixes issue 1 by hiding the Region column on invoices before DC-specific pricing was implemented. This month's invoice and invoices going forward should include region data for invoice items.
- Fixes issue 2 by displaying monthly prices returned from the API as decimals (at this point, only Jakarta and Sao Paulo) with two decimal places. Monthly prices that are integers (e.g. a 2GB linode at $10/month) should remain integers (not $10.00). We want to avoid adding extra zeroes on the screen (making the price look higher than it is) if it is not necessary. For regions with decimals in their monthly prices, we display to two decimal place rather than rounding or truncating to an integer because prices should be consistent with the prices listed in the linode.com docs.
- Updates unit and cypress tests accordingly for both these fixes.

## Preview 📷
| Before  | After   |
| ------- | ------- |
| ![Screenshot 2023-10-04 at 7 46 52 PM](https://github.com/linode/manager/assets/114685994/227479c2-17db-4258-888c-2ea0ba9c59a1) | ![Screenshot 2023-10-04 at 7 46 25 PM](https://github.com/linode/manager/assets/114685994/ddaa99a0-4ed9-4276-8c19-ec6054259723) |
| ![Screenshot 2023-10-04 at 7 47 13 PM](https://github.com/linode/manager/assets/114685994/3f453afb-04bb-45c2-ac1b-84b5f8ae4e29) | ![Screenshot 2023-10-04 at 7 48 21 PM](https://github.com/linode/manager/assets/114685994/1cb7edf0-5a9b-4ddc-8e42-be79f282bfe7) |
| ![Screenshot 2023-10-04 at 9 55 38 PM](https://github.com/linode/manager/assets/114685994/2f571ed6-62f7-4363-991a-f43f5f86ab26) | ![Screenshot 2023-10-04 at 7 48 32 PM](https://github.com/linode/manager/assets/114685994/b259b7d6-c926-4721-a964-e10d9cdb8bac) |
| ![Screenshot 2023-10-04 at 10 00 23 PM](https://github.com/linode/manager/assets/114685994/20b48d6d-db17-4e8a-a88e-1e60b482b959) | ![Screenshot 2023-10-04 at 7 49 21 PM](https://github.com/linode/manager/assets/114685994/8ad8baf0-b89b-419e-bce1-7d5b074b6ab1) |
| ![Screenshot 2023-10-04 at 9 59 56 PM](https://github.com/linode/manager/assets/114685994/46a6a33c-880b-4de3-bc86-0b73d11a214d) | ![Screenshot 2023-10-04 at 7 49 41 PM](https://github.com/linode/manager/assets/114685994/8fc102c3-aaa7-4c68-8d64-e26f878e0c30) |
![Screenshot 2023-10-04 at 7 55 45 PM](https://github.com/linode/manager/assets/114685994/9cd78d5f-2ac6-40e2-a06f-f223e65a60ae) | ![Screenshot 2023-10-04 at 7 55 29 PM](https://github.com/linode/manager/assets/114685994/03ae3e6c-7931-448f-8da9-36128729366f) |
|![Screenshot 2023-10-04 at 10 05 09 PM](https://github.com/linode/manager/assets/114685994/294fc567-7e69-42c8-98fa-fedf4f224183) | ![Screenshot 2023-10-04 at 10 04 48 PM](https://github.com/linode/manager/assets/114685994/05c1133a-18cd-41be-b3a9-2a82c03dc44b) |


## How to test 🧪
1. **How to setup test environment?**
- Check out this PR.
- Have another tab open to compare to prod.
2. **How to reproduce the issue (if applicable)?**
- In prod, through the linode create flow, kube create flow, and kube node pools add/resize pool flows.
- Observe you see monthly prices in the plans tables, summaries, and drawers with a weird number of decimal points.
- Go to your latest account invoice on the billing page.
- Observe you see a regions column but no regions listed.
3. **How to verify changes?**
- On localhost, go through the linode create flow, kube create flow, and kube node pools add/resize pool flows.
- Observe you see monthly prices in the plans tables, summaries, and drawers are unchanged (no decimals) for resources outside of Jakarta or Sao Paulo.
- Observe you see monthly prices in the plans tables, summaries, and drawers with two decimals for resources in Jakarta or Sao Paulo, which have DC-specific pricing.
- Go to your latest account invoice on the billing page.
- Observe you do not see a regions column any of the following: details page (invoice table), csv download, pdf download.
- Edit the regions mock data in serverHandler.ts or the factory to include an invoice with a future date.
- Observe that you do see the regions column in all of the following: details page (invoice table), csv download, pdf 
- If you want to confirm regions will show on a future invoice (like the end of this month) update the existing mock request to include a date of `2023-10-31T18:04:01` and the go to http://localhost:3000/account/billing/invoices/1234:
```
  rest.get('*/account/invoices/:invoiceId', (req, res, ctx) => {
    const linodeInvoice = invoiceFactory.build({
      date: '2022-12-01T18:04:01',
      id: 1234,
      label: 'LinodeInvoice',
    });
    return res(ctx.json(linodeInvoice));
  }),
```

4. **How to run Unit or E2E tests?**
```
yarn test
```
e2es should pass, but I've been seeing issues with the LKE smoke tests locally due to some odd behavior with the region select scrolling out of view and returning `no results` after a selection is made -- that doesn't replicate in the UI, so not sure what is going on here.
